src/ParallelAlgorithms/Events/ObserveNorms.hpp

```diff
@@ -321,6 +330,55 @@ ObserveNorms<tmpl::list<ObservableTensorTags...>,
   }
 }
 
+// implementation of ObserveNorms::operator() factored out to save on compile
+// time and compile memory
+namespace ObserveNorms_impl {
+void check_norm_is_observable(const std::string& tensor_name,
+                              bool tag_has_value);
+
+template <size_t Dim>
+void fill_norm_values_and_names(
+    gsl::not_null<std::unordered_map<
+        std::string, std::pair<std::vector<double>, std::vector<std::string>>>*>
+        norm_values_and_names,
+    const std::pair<std::vector<std::string>, std::vector<DataVector>>&
+        names_and_components,
+    const Mesh<Dim>& mesh, const DataVector& det_jacobian,
+    const std::string& tensor_name, const std::string& tensor_norm_type,
+    const std::string& tensor_component, size_t number_of_points);
+}  // namespace ObserveNorms_impl
+
+template <typename... ObservableTensorTags, typename... NonTensorComputeTags,
+          typename ArraySectionIdTag>
+template <typename TensorToObserveTag, typename ComputeTagsList,
+          typename DataBoxType, size_t Dim>
+void ObserveNorms<tmpl::list<ObservableTensorTags...>,
+                  tmpl::list<NonTensorComputeTags...>, ArraySectionIdTag>::
+    observe_norms_impl(
+        const gsl::not_null<std::unordered_map<
+            std::string,
+            std::pair<std::vector<double>, std::vector<std::string>>>*>
+            norm_values_and_names,
+        const ObservationBox<ComputeTagsList, DataBoxType>& box,
+        const Mesh<Dim>& mesh, const DataVector& det_jacobian,
+        const size_t number_of_points) const {
+  const std::string tensor_name = db::tag_name<TensorToObserveTag>();
+  // Loop over ObservableTensorTags and see if it was requested to be observed.
+  // This approach allows us to delay evaluating any compute tags until they're
+  // actually needed for observing.
```

Comment:
Regarding the lambda, I pulled it out because it was a big part of the savings for the `ObserveNorms` chhanges, though the total savings weren't huge. With the current change I have (pulling out the `tmpl::for_each` into a function AND factoring out `check_norm_is_observable` and `fill_norm_values_and_names`), it saved 0.21 GB and 24.5s for clang-13 Release and 0.17GB and 27.9s for gcc-9 Release. Now if I put the `tmpl::for_each` back in (but still keep `check_norm_is_observable` and `fill_norm_values_and_names` factored out), it adds back 0.08GB (of 0.21GB) and 13s (of 24.5s) for clang and 0.12 GB (of 0.17GB) and 14s (of 27.9s) for gcc. (note: The timing may have some noise, I didn't run it multiple times to get an average)

Let me know how you feel about the tradeoff, but I will move the comment back either way!

Reply:
Keep as-is. the `tmpl::for_each` would be a 50%+ regression, which is huge.

Reply:
Yep. Sounds like `tmpl::for_each` is a lot more expensive than a parameter pack expansion :(

Reply:
Sounds good. I moved the comment back down